Compiling issue on Xcode 15.0 + conda (macOS)

<!-- Please fill in the following information to report a problem with Cantera. If you have a question about using Cantera, please post it on our Google Users' Group (https://groups.google.com/forum/#!forum/cantera-users). Feature enhancements should be discussed in the dedicated Cantera enhancements repository (https://github.com/Cantera/enhancements/new/choose) -->

**Problem description**

<!-- A clear and concise description of what the bug is. -->
Compilation on Xcode 15.0 fails. 

**Steps to reproduce**

<!-- A minimal, complete, and reproducible example demonstrating the problem. See https://stackoverflow.com/help/minimal-reproducible-example for additional suggestions on how to create such an example. -->

1. Install conda via `brew install miniforge` (or `brew install miniconda`)
2. Create conda environment according to [instructions](https://cantera.org/compiling/compilation-reqs.html#conda-requirements)
3. Activate environment & run `scons build`
4. Compilation fails.

**Behavior**

On Ventura:
```
dyld[42265]: Symbol not found: __ZNKSt3__16locale9use_facetERNS0_2idE
  Referenced from: <36C0E55E-CB4E-3D7F-BA2E-789336B4709B> /Volumes/Data/work/GitHub/cantera/.sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055
  Expected in:     <no uuid> unknown
sh: line 1: 42265 Abort trap: 6           .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055 > ".sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055.out"
ERROR: The C++ compiler is not correctly configured (failed at linking stage).
```
<details>
<pre><code>% scons build
scons: Reading SConscript files ...
SCons 4.5.2 is using the following Python interpreter:
    /opt/homebrew/Caskroom/miniforge/base/envs/cantera-dev/bin/python (Python 3.9)
INFO: Compiling on 'Apple M2'
INFO: Building Cantera from git commit '9696e27ea'
INFO: Configuration variables read from 'cantera.conf' and command line:

INFO: Adding conda include and library paths: /opt/homebrew/Caskroom/miniforge/base/envs/cantera-dev
Checking for C++ header file cmath... yes
dyld[42265]: Symbol not found: __ZNKSt3__16locale9use_facetERNS0_2idE
  Referenced from: <36C0E55E-CB4E-3D7F-BA2E-789336B4709B> /Volumes/Data/work/GitHub/cantera/.sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055
  Expected in:     <no uuid> unknown
sh: line 1: 42265 Abort trap: 6           .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055 > ".sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055.out"
ERROR: The C++ compiler is not correctly configured (failed at linking stage).

See 'config.log' for details.
</code></pre>
where `config.log` contains
<pre><code>% cat config.log
file /Volumes/Data/work/GitHub/cantera/SConstruct,line 1128:
	Configure(confdir = .sconf_temp)
scons: Configure: Checking for C++ header file cmath...
.sconf_temp/conftest_4501b9bf892a8ba32d6185db437667f6_0.cpp <-
  |
  |#include <cmath>
  |
  |
g++ -o .sconf_temp/conftest_4501b9bf892a8ba32d6185db437667f6_0_bfdf4a7d91b3eddd0f366c226d535e62.o -c -isystem /opt/homebrew/Caskroom/miniforge/base/envs/cantera-dev/include -std=c++17 -O3 -Wno-inline -g -DNDEBUG .sconf_temp/conftest_4501b9bf892a8ba32d6185db437667f6_0.cpp
scons: Configure: yes

.sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0.cpp <-
  |#include <cmath>
  |#define Q(x) #x
  |#define QUOTE(x) Q(x)
  |#include <iostream>
  |int main(int argc, char** argv) {
  |    std::cout << cos(0. * argc) << std::endl;
  |    return 0;
  |}
  |
g++ -o .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0.o -c -isystem /opt/homebrew/Caskroom/miniforge/base/envs/cantera-dev/include -std=c++17 -O3 -Wno-inline -g -DNDEBUG .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0.cpp
g++ -o .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055 .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0.o -L/opt/homebrew/Caskroom/miniforge/base/envs/cantera-dev/lib
.sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055 > .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055.out
</code></pre>
</details>

On Sonoma:
```
dyld[5849]: Library not loaded: @rpath/libc++.1.dylib
  Referenced from: <0DBF2E3D-9611-3C0F-AADA-59DE79DFF77F> /Volumes/Data/work/GitHub/cantera/.sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055
  Reason: no LC_RPATH's found
sh: line 1:  5849 Abort trap: 6           .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055 > ".sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055.out"
ERROR: The C++ compiler is not correctly configured (failed at linking stage).
```
<details>
<pre><code>% scons build -j4
scons: Reading SConscript files ...
SCons 4.5.1 is using the following Python interpreter:
    /opt/homebrew/Caskroom/miniforge/base/envs/cantera-dev/bin/python (Python 3.10)
INFO: Compiling on 'Apple M2 Pro'
INFO: Building Cantera from git commit '9696e27ea'
INFO: Configuration variables read from 'cantera.conf' and command line:

INFO: Adding conda include and library paths: /opt/homebrew/Caskroom/miniforge/base/envs/cantera-dev
Checking for C++ header file cmath... yes
dyld[5849]: Library not loaded: @rpath/libc++.1.dylib
  Referenced from: <0DBF2E3D-9611-3C0F-AADA-59DE79DFF77F> /Volumes/Data/work/GitHub/cantera/.sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055
  Reason: no LC_RPATH's found
sh: line 1:  5849 Abort trap: 6           .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055 > ".sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055.out"
ERROR: The C++ compiler is not correctly configured (failed at linking stage).

See 'config.log' for details
</code></pre>
where `config.log` contains
<pre><code>% cat config.log
file /Volumes/Data/work/GitHub/cantera/SConstruct,line 1128:
	Configure(confdir = .sconf_temp)
scons: Configure: Checking for C++ header file cmath...
.sconf_temp/conftest_4501b9bf892a8ba32d6185db437667f6_0.cpp <-
  |
  |#include <cmath>
  |
  |
g++ -o .sconf_temp/conftest_4501b9bf892a8ba32d6185db437667f6_0_bfdf4a7d91b3eddd0f366c226d535e62.o -c -isystem /opt/homebrew/Caskroom/miniforge/base/envs/cantera-dev/include -std=c++17 -O3 -Wno-inline -g -DNDEBUG .sconf_temp/conftest_4501b9bf892a8ba32d6185db437667f6_0.cpp
scons: Configure: yes

.sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0.cpp <-
  |#include <cmath>
  |#define Q(x) #x
  |#define QUOTE(x) Q(x)
  |#include <iostream>
  |int main(int argc, char** argv) {
  |    std::cout << cos(0. * argc) << std::endl;
  |    return 0;
  |}
  |
g++ -o .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0.o -c -isystem /opt/homebrew/Caskroom/miniforge/base/envs/cantera-dev/include -std=c++17 -O3 -Wno-inline -g -DNDEBUG .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0.cpp
g++ -o .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055 .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0.o -L/opt/homebrew/Caskroom/miniforge/base/envs/cantera-dev/lib 
.sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055 > .sconf_temp/conftest_a4c09ef5b1a9d7f569d7b807c37828ab_0_ed8c1642fe6cca76c1ce173236e40055.out
</code></pre>
</details>

**System information**

- Cantera version: `main`
- OS: macOS Ventura / Sonoma (on Apple Silicon M2)
- Python/MATLAB/other software versions: conda 23.3.1 (via `brew install miniforge`) + Apple clang version 15.0.0 (clang-1500.0.40.1), xcode-select version 2397 (CLI tools only)

**Other thoughts**

This is presumably a conda issue that may have to be resolved upstream. At the same time, compilations still work for other projects.

@speth  I tried it, but get an error. What I did was create a new conda environment, install all packages - clangxx, cython, scons, humpy and hdf5 via conda (clangxx via conda-forge as you specified). I had to specify the new compiler location as well has include paths via `CPATH` and `LD_LIBRARY_PATH`. See the config.log below 

```
cat config.log 
file /Users/santosh1/packages/canteraSource/SConstruct,line 1128:
	Configure(confdir = .sconf_temp)
scons: Configure: Checking for C++ header file cmath... 
scons: Configure: ".sconf_temp/conftest_0.cpp" is up to date.
scons: Configure: The original builder output was:
  |.sconf_temp/conftest_0.cpp <-
  |  |
  |  |#include <cmath>
  |  |
  |  |
  |
/Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/clang++ -o .sconf_temp/conftest_0.o -c -isystem /Users/santosh1/opt/anaconda3/envs/py311-cxx/include -isystem /Users/santosh1/packages/boost_1_83_0 -std=c++17 -fcolor-diagnostics -O3 -g -DNDEBUG .sconf_temp/conftest_0.cpp
In file included from .sconf_temp/conftest_0.cpp:2:
In file included from /Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/cmath:317:
In file included from /Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/math.h:316:
/Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/stdlib.h:150:34: error: unknown type name 'ldiv_t'
inline _LIBCPP_INLINE_VISIBILITY ldiv_t div(long __x, long __y) _NOEXCEPT {
                                 ^
/Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/stdlib.h:151:12: error: no member named 'ldiv' in the global namespace
  return ::ldiv(__x, __y);
         ~~^
/Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/stdlib.h:154:34: error: unknown type name 'lldiv_t'
inline _LIBCPP_INLINE_VISIBILITY lldiv_t div(long long __x,
                                 ^
/Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/stdlib.h:156:12: error: no member named 'lldiv' in the global namespace
  return ::lldiv(__x, __y);
         ~~^
In file included from .sconf_temp/conftest_0.cpp:2:
In file included from /Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/cmath:317:
/Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/math.h:388:31: error: use of undeclared identifier 'FP_NAN'
  return __builtin_fpclassify(FP_NAN, FP_INFINITE, FP_NORMAL, FP_SUBNORMAL, FP_ZERO, __x);
                              ^
/Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/math.h:388:39: error: use of undeclared identifier 'FP_INFINITE'
  return __builtin_fpclassify(FP_NAN, FP_INFINITE, FP_NORMAL, FP_SUBNORMAL, FP_ZERO, __x);
                                      ^
/Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/math.h:388:52: error: use of undeclared identifier 'FP_NORMAL'
  return __builtin_fpclassify(FP_NAN, FP_INFINITE, FP_NORMAL, FP_SUBNORMAL, FP_ZERO, __x);
                                                   ^
/Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/math.h:388:63: error: use of undeclared identifier 'FP_SUBNORMAL'
  return __builtin_fpclassify(FP_NAN, FP_INFINITE, FP_NORMAL, FP_SUBNORMAL, FP_ZERO, __x);
                                                              ^
/Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/math.h:388:77: error: use of undeclared identifier 'FP_ZERO'
  return __builtin_fpclassify(FP_NAN, FP_INFINITE, FP_NORMAL, FP_SUBNORMAL, FP_ZERO, __x);
                                                                            ^
/Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/math.h:393:21: error: use of undeclared identifier 'FP_ZERO'
  return __x == 0 ? FP_ZERO : FP_NORMAL;
                    ^
/Users/santosh1/opt/anaconda3/envs/py311-cxx/bin/../include/c++/v1/math.h:393:31: error: use of undeclared identifier 'FP_NORMAL'
  return __x == 0 ? FP_ZERO : FP_NORMAL;
                              ^
11 errors generated.
scons: Configure: no
```

--- a/SConstruct
+++ b/SConstruct
@@ -1199,6 +1199,19 @@ def get_expression_value(includes, expression, defines=()):
               '}\n'))
     return '\n'.join(s)
 
+# Need to ensure that RPATH is working before we can continue with library checks
+env['HAS_CLANG'] = conf.CheckDeclaration('__clang__', '', 'C++')
+
+if env['OS'] == 'Solaris' or env['HAS_CLANG']:
+    env["RPATHPREFIX"] = "-Wl,-rpath,"
+
+# Add initial libraries to RPATH
+if env["OS"] == "Darwin" and env["use_rpath_linkage"] and not env.subst("$__RPATH"):
+    # SCons doesn't want to specify RPATH on macOS, so circumvent that behavior by
+    # specifying this directly as part of LINKFLAGS
+    env.Append(LINKFLAGS=[env.subst(f'$RPATHPREFIX{x}$RPATHSUFFIX')
+                          for x in env['RPATH']])
+
 # Check that libraries link correctly
 cmath_check_source = get_expression_value(["<cmath>"], "cos(0. * argc)")
 retcode, cmath_works = conf.TryRun(cmath_check_source, ".cpp")
@@ -1343,7 +1356,6 @@ elif conf.CheckDeclaration('_LIBCPP_VERSION', '#include <iostream>', 'C++'):
 else:
     env['cxx_stdlib'] = []
 
-env['HAS_CLANG'] = conf.CheckDeclaration('__clang__', '', 'C++')
 if not env["using_apple_clang"]:
     # This checks for these three libraries in order and stops when it finds the
     # first success. Intel = iomp5, LLVM/clang = omp, GCC = gomp. Since gomp is
@@ -1909,15 +1921,12 @@ def cdefine(definevar, configvar, comp=True, value=1):
     else:
         configh[definevar] = None
 
-if env['OS'] == 'Solaris' or env['HAS_CLANG']:
-    env["RPATHPREFIX"] = "-Wl,-rpath,"
-
+# Update RPATH with additional library directories
 if env["OS"] == "Darwin" and env["use_rpath_linkage"] and not env.subst("$__RPATH"):
     # SCons doesn't want to specify RPATH on macOS, so circumvent that behavior by
     # specifying this directly as part of LINKFLAGS
-    env.Append(LINKFLAGS=[env.subst(f'$RPATHPREFIX{x}$RPATHSUFFIX')
-                          for x in env['RPATH']])
-
+    env.Append(LINKFLAGS=[d for x in env['RPATH']
+        if (d := env.subst(f'$RPATHPREFIX{x}$RPATHSUFFIX')) not in env['LINKFLAGS']])
 
 if env.get('has_sundials_lapack') and env['use_lapack']:
     configh['CT_SUNDIALS_USE_LAPACK'] = 1

